### PR TITLE
Recipe logic fix/clean-ups

### PR DESCRIPTION
**Pull request by PrototypeTrousers**
**What:**
This is an attempt to fix a issue with multiblocks apparently stopping running recipes with ingredients enough for a recipe to run idling on their input busses

**Implementation Details:**
Made the forceRecipeRecheck in AbstractRecipeLogic also reset the previouly cached recipe
Fix busses not removing the previous controller object properly on multiblock invalidatio (memory leak)
Extra null checks to avoid null entries in the notifiables list

**Outcome:**
Hopefully fixes the issues reported. *crosses fingers,hands,arms and legs*

### CODE REVIEW

File: src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java [hunk 1]
```diff
@@ -103,6 +103,25 @@ protected IMultipleTankHandler getOutputTank() {
...
     }
 
+    @Override
+    protected boolean canWorkWithInputs() {
+        MultiblockWithDisplayBase controller = (MultiblockWithDisplayBase) metaTileEntity;
+        if (controller instanceof RecipeMapMultiblockController) {
+            RecipeMapMultiblockController distinctController = (RecipeMapMultiblockController) controller;
+
+            if (distinctController.canBeDistinct() && distinctController.isDistinct()) {
+                if (invalidatedInputList.isEmpty() || !metaTileEntity.getNotifiedFluidInputList().isEmpty()) return true;
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java:113]: Should this check `invalidatedInputList.isEmpty() && !metaTileEntity.getNotifiedItemInputList.isEmpty() ||...` instead of just checking if we have no invalidated any inputs previously?

File: src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java [hunk 2]
```diff
@@ -103,6 +103,25 @@ protected IMultipleTankHandler getOutputTank() {
...
+                for (IItemHandlerModifiable bus : metaTileEntity.getNotifiedItemInputList()) {
+                    if (invalidatedInputList.contains(bus)) {
+                        return true;
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java:116]: As well as returning true, should we remove the selected bus from the `invalidatedInputList`?

In fact, we might want to delay returning true until we finish looping over all the buses, so that we can remove any that were notified from `invalidatedInputList` and then only return true if there were buses that were removed from the list.

File: src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java (hunk 1 above)
Comment (PrototypeTrousers) [reply src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java]: we only care if the inputs changed if it is currently invalidated

File: src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java [hunk 3]
```diff
@@ -103,6 +104,39 @@ protected IMultipleTankHandler getOutputTank() {
...
     }
 
+    @Override
+    protected boolean canWorkWithInputs() {
+        MultiblockWithDisplayBase controller = (MultiblockWithDisplayBase) metaTileEntity;
+        if (controller instanceof RecipeMapMultiblockController) {
+            RecipeMapMultiblockController distinctController = (RecipeMapMultiblockController) controller;
+
+            if (distinctController.canBeDistinct() && distinctController.isDistinct()) {
+                boolean canWork = false;
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java:114]: Do we want to add a check here like
```java
if(!hasNotifiedInputs()){
        return false;
        }
```
to match the case in the `super` method and for short circuiting?

As if we do not have any notified inputs, then there is nothing for the multiblock to work with, and we have not checked if we have notified inputs yet.

File: src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java [hunk 4]
```diff
@@ -168,10 +190,8 @@ protected void trySearchNewRecipeDistinct() {
         for (int i = 0; i < importInventory.size(); i++) {
             IItemHandlerModifiable bus = importInventory.get(i);
             // Skip this bus if no recipe was found last time and the inventory did not change
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java:192]: Could you adjust the comment on this line, since the `if` check was updated.

File: src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java [hunk 5]
```diff
@@ -103,6 +104,39 @@ protected IMultipleTankHandler getOutputTank() {
...
+                    metaTileEntity.getNotifiedFluidInputList().clear();
+                }
+                Iterator<IItemHandlerModifiable> iterator = metaTileEntity.getNotifiedItemInputList().iterator();
+                while (iterator.hasNext()) {
+                    IItemHandlerModifiable bus = iterator.next();
+                    if (invalidatedInputList.remove(bus)) {
+                        canWork = true;
+                    }
+                    iterator.remove();
+                }
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java:130]: This section of code could be put into an `else` for the `if` statement check on the notified fluids. Since if we have notified fluids, we already clear the invalidated list, meaning that the check in the while loop will never be true.

File: src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java (hunk 3 above)
Comment (PrototypeTrousers) [reply src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java:114]: we clear the notified status after a recipe is run, so short circuiting it here would mean we run 1 recipe and stall till the input changes again

File: src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java [hunk 6]
```diff
@@ -103,6 +104,43 @@ protected IMultipleTankHandler getOutputTank() {
...
 
+    @Override
+    protected boolean canWorkWithInputs() {
+        MultiblockWithDisplayBase controller = (MultiblockWithDisplayBase) metaTileEntity;
+        if (controller instanceof RecipeMapMultiblockController) {
+            RecipeMapMultiblockController distinctController = (RecipeMapMultiblockController) controller;
+
+            if (distinctController.canBeDistinct() && distinctController.isDistinct()) {
+                boolean canWork = false;
+                if (!hasNotifiedInputs() && !hasNotifiedOutputs()) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java:115]: Since this is in `canWorkWithInputs`, checking if the outputs are notified is probably not correct. That is done in `canFitNewOutputs`.

**Comment by ALongStringOfNumbers** [comment]
I meant remove the `!hasNotifiedOutputs()` part of the short circuit check, instead of removing the entire short circuit check.

`hasNotifiedOutputs` is checked in `canFitNewOutputs`, which is checked in conjunction with `canWorkWithInputs`, so the additional check here for if the outputs are notified is redundant. Plus, it does not make sense to reject the inputs because the outputs were not notified.

So basically I just wanted
```java
if (!hasNotifiedInputs()) {
    return false;
}
```
instead of
```java
if (!hasNotifiedInputs() && !hasNotifiedOutputs()) {
    return false;
}
```

File: src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java [hunk 7]
```diff
@@ -103,6 +104,40 @@ protected IMultipleTankHandler getOutputTank() {
...
+
+            if (distinctController.canBeDistinct() && distinctController.isDistinct()) {
+                boolean canWork = false;
+                if (invalidatedInputList.isEmpty()) {
+                    return true;
+                }
+                if (!metaTileEntity.getNotifiedFluidInputList().isEmpty()) {
+                    canWork = true;
+                    invalidatedInputList.clear();
+                    metaTileEntity.getNotifiedFluidInputList().clear();
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java:121]: There was a slight logic change that I missed when I asked to introduce the `else` part of this check. That is that the notified item inputs are no longer always cleared, but are instead only cleared if there are no notified fluid inputs.

This logic change can be compensated for by clearing the notified Item Inputs here, or just removing the `else` part of this statement that was introduced, and move back to always creating the iterator. I should not review when I am tired.

**Review by ALongStringOfNumbers — APPROVED** [review]
